### PR TITLE
feat: Implement drag & drop for metrics and filters labels

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
We can rearrange metrics and filters labels with drag&drop functionality.

We use very old version of react-dnd (2.5.4) - it might be worth to consider an upgrade. I had to disable some lint and ts errors - e.g. due to type definitions for typescript.
react-dnd v2.5.4 has a bug that prevents using refs in functional components - for that reason I had to use `findDOMNode`, which should be replaced as soon as we do the upgrade.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/15073128/102925235-4b59f500-4493-11eb-8584-76ed087f0103.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC: @junlincc @villebro @rusackas @adam-stasiak

### CODE REVIEW

**Comment by adam-stasiak** [comment]
Tested - No issues from me.

File: superset-frontend/src/explore/components/controls/AdhocFilterControl.jsx [hunk 1]
```diff
@@ -39,6 +38,7 @@ import {
 } from '../OptionControls';
 import Icon from '../../../components/Icon';
 import AdhocFilterPopoverTrigger from '../AdhocFilterPopoverTrigger';
+import DndWithHTML5Backend from '../../DndContextProvider';
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/AdhocFilterControl.jsx:41]:
```suggestion
import DndWithHTML5Backend from 'src/explore/DndContextProvider';
```

Nit: replace this (and maybe other imports in this file, too) with absolute path.

File: superset-frontend/src/explore/components/controls/MetricsControl.jsx [hunk 2]
```diff
@@ -238,6 +242,17 @@ class MetricsControl extends React.PureComponent {
...
+  moveLabel(dragIndex, hoverIndex) {
+    const { value } = this.state;
+
+    const newValues = [...value];
+    [newValues[hoverIndex], newValues[dragIndex]] = [
+      newValues[dragIndex],
+      newValues[hoverIndex],
+    ];
+    this.setState({ value: newValues });
+  }
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/MetricsControl.jsx:254]: Not a blocker, but maybe we can abstract `AdhocFilterControl` and `MetricsControl` into a shared `AdhocOptionsControl` (hope someone can come up a better name than me).

File: superset-frontend/src/explore/components/OptionControls.tsx [hunk 3]
```diff
@@ -17,12 +17,28 @@
...
+// TODO: remove ts-ignore after we upgrade react-dnd
+// @ts-ignore
+import { DragSource, DropTarget } from 'react-dnd';
 import { styled, useTheme } from '@superset-ui/core';
 import { ColumnOption } from '@superset-ui/chart-controls';
 import Icon from '../../components/Icon';
 import { savedMetricType } from '../types';
 
-const OptionControlContainer = styled.div<{ isAdhoc?: boolean }>`
+const TYPE = 'label-dnd';
```
Comment (ktmud) [inline superset-frontend/src/explore/components/OptionControls.tsx:30]:
```suggestion
const DND_ITEM_TYPE = 'label-dnd';
```

**Review by ktmud — COMMENTED** [review]
Thanks for prioritizing this feature! I think this works. Added a couple of nits. Hope we can further iterate on this and use placeholders or DnD indicators, instead of directly switch orders when hover, to make the experiencer more smooth.

While testing, I also noticed that almost all controls now triggers a new query by default. Seems like a separate bug, so stamping in favor of faster iteration.

**Review by ktmud — APPROVED** [review]

**Pull request merged**